samples/bookinfo/src/productpage/productpage.py

```diff
@@ -80,18 +87,69 @@
     "reviews" : reviews,
 }
 
+# A very basic OpenTracing tracer (with null reporter)
+tracer = Tracer(
+    one_span_per_rpc=True,
+    service_name='productpage',
+    reporter=NullReporter(),
+    sampler=ConstSampler(decision=True),
+    extra_codecs={Format.HTTP_HEADERS: B3Codec()}
+)
+
+
+def trace():
+    '''
+    Function decorator that creates opentracing span from incoming b3 headers
+    '''
+    def decorator(f):
+        def wrapper(*args, **kwargs):
+            request = stack.top.request
+            try:
+                # Create a new span context, reading in values (traceid,
+                # spanid, etc) from the incoming x-b3-*** headers.
+                span_ctx = tracer.extract(
+                    Format.HTTP_HEADERS,
+                    dict(request.headers)
+                )
+                # Note: this tag means that the span will *not* be
+                # a child span. It will use the incoming traceid and
+                # spanid. We do this to propagate the headers verbatim.
+                rpc_tag = {tags.SPAN_KIND: tags.SPAN_KIND_RPC_SERVER}
+                span = tracer.start_span(
+                    operation_name='op', child_of=span_ctx, tags=rpc_tag
+                )
+            except Exception as e:
+                # We failed to create a context, possibly due to no
+                # incoming x-b3-*** headers. Start a fresh span.
+                # Note: This is a fallback only, and will create fresh headers,
+                # not propagate headers.
+                span = tracer.start_span('op')
+            with span_in_context(span):
+                r = f(*args, **kwargs)
+                return r
+        wrapper.__name__ = f.__name__
+        return wrapper
+    return decorator
+
+
 def getForwardHeaders(request):
     headers = {}
 
+    # x-b3-*** headers can be populated using the opentracing span
+    span = get_current_span()
+    carrier = {}
+    tracer.inject(
+        span_context=span.context,
+        format=Format.HTTP_HEADERS,
+        carrier=carrier)
+
+    headers.update(carrier)
+
+    # We handle other (non x-b3-***) headers manually
     if 'user' in session:
         headers['end-user'] = session['user']
 
     incoming_headers = [ 'x-request-id',
-                         'x-b3-traceid',
-                         'x-b3-spanid',
-                         'x-b3-parentspanid',
-                         'x-b3-sampled',
-                         'x-b3-flags',
                          'x-ot-span-context'
```

Comment:
We still need this right? Otherwise, stuff didn't work with Envoy. If this is not needed, please remove it.

Reply:
`x-ot-span-context` should no longer be required as the `x-b3-*` are being propagated.

Reply:
I left things unchanged initially, just to reduce chances of unexpected side effects.

Once we're confident we don't need x-ot-span-context we could remove it in a later patch (for all services).

If folks think we should just take it out now, let me know.

Reply:
Each update to bookinfo involves validating e2e tests, updating the official images, and another pr to update image versions on all yamls, etc. all manual. So let’s just do all these changes in same pr.